Add tool to display emergency log message full-sceen on boot failure.

1awesomeJ opened the pull request:
<!-- devel-freezer = {"comment-id":"1624712630","freezing-tag":"v254-rc3"} -->

1awesomeJ commented:
I want to start with a function that fetches the oldest message in the journal and displays it to the console.
Currently I set PRIORITY to 5(notice) to see if I'll get any messages. I'm currently not getting any.

1awesomeJ commented:
@bluca 
Following Poettering's guidance, I created this bsod tool.
Currently it just tries to fetch the oldest message from the current boot with a log level of "emergency", and prints it to a console. 

Subsequently, I have to modify it to take over the entire screen, turn it blue, and display the QR code.

Howevever, I am currently not getting an output at all, even when i change the PRIORITY to numbers other than 0.
What am I doing wrong sir?

Thank you.

poettering commented on man/rules/meson.build:
```diff
@@ -898,6 +898,7 @@ manpages = [
   ''],
```
something is wrong in your git setup. you are using your name where you should use your email address, see the git author info for this commit

poettering commented on man/systemd-bsod.xml:
```diff
@@ -0,0 +1,62 @@
+
+  <refmeta>
+    <refentrytitle>systemd-bsod</refentrytitle>
+    <manvolnum>1</manvolnum>
```
please move to section "8", which is where we put all tools users aren't supposed to call directly from the cmdline

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+               program_invocation_short_name);
+}
+
+static char* first_emerg_boot_message(void) {
```
please return the error code here as we usually do, i.e. change the prototype to:

```c
static int acquire_first_emergency_log_message(char **ret) {
```

or so

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+                return r;
+
+        _cleanup_close_ int fd = -EBADF;
+        char * message = first_emerg_boot_message();
```
this can fail, but there's no error handling

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+        _cleanup_close_ int fd = -EBADF;
+        char * message = first_emerg_boot_message();
+
+        fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
```
so for this I think we should not go to /dev/console (which is a magic device that basically points to wherever boot time logs and stuff should go to which can be serial or similar) but the VT subsystem (which is the virtual terminal, i.e. a textual interface to a local physical display device, i.e. no serial), and I think we should not interfere with the usual log/progress output that /dev/console gets. Hence I'd suggest we allocate a full new VT for this, and then switch to that.

(in case you wonder what a VT is, it's this archaic textual display logic that the linux kernel uses to do early boot logging before wayland/x11 take over, and that you can log into via Alt-F2, Alt-F3, …)

i.e. open /dev/tty1 temporarily, then issue the VT_GETSTATE ioctl call on it which tells you which VTs are currently allocated via a bitmask. Look for the first free VT (i.e. determine lowest unset bit, then format /dev/tty%i with it plus one. Then open that, and use that. switch to it via the VT_ACTIVATE ioctl, then clear it by output ANSI_HOME_CLEAR on it, then display the message there.

If this sounds like a bit much, grep our sources for VT_GETSTAT, we already call that ioctl elsewhere for other reasons.

halfline replied:
`VT_OPENQRY` might be easier than `VT_GETSTATE` and `__builtin_ctz(~vt_stat.v_state)` or whatever.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+        }
+        sd_id128_to_string(boot_id, boot_id_string);
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=%s", boot_id_string);
```
use SD_ID128_FORMAT_STR() and SD_ID128_FORMAT_VAL() instead of calling sd_id128_to_string() directly

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
```
maybe add here (mostly as decoration for those who might not be aware that 0 is LOG_EMERG):
```c
assert_cc(0 == LOG_EMERG);
```

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+                r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read journal message: %m");
+                        continue;
```
if you fail to read the journal you should give up

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,143 @@
+                goto clean_journal;
+        }
+
+        SD_JOURNAL_FOREACH(j) {
```
don't use a loop if you just want the first message. Use sd_journal_seek_first() once, plus one sd_journal_next() instead

poettering commented on man/systemd-bsod.xml:
```diff
@@ -0,0 +1,62 @@
+
+  <refnamediv>
+    <refname>systemd-bsod</refname>
+    <refpurpose>Displays the systemd blue screen of death when booting fails</refpurpose>
```
let's maybe not use the word "blue screen of death" in the text here. calling the service "bsod" is a nod towards the direction of windows, but we should leave it like that. In the description let's just say "Displays boot-time emergency log message full-screen" or so.

1awesomeJ replied:
Wow, Thank you so much @poettering for all the detailed reviews.
I'm really thankful.
I figured I should just push what I have. I counted on getting these reviews, to help me get the patch super refined.
Thank you sso much sir.

Every single point very duly noted.
It's a privilege.

1awesomeJ commented:
I'm read that the APIs return errno-style errors of failure. Kindly review my choice of error types sir.

1awesomeJ commented:
I looked up the format string and format value macros sir, and I've now used them in the patch.

So it's VT I need to look up now.

perhaps I have too many log messages?

1awesomeJ commented:
Still I don't get any messages from the journal.
```journalctl -b -p debug _UID=0 -o verbose -n1``` shows an output.

But running the execuatble keeps saying "No entries in the journal"

I have also gradually reduced the match filters from 3 to 0.

perhaps there's something else I'm missing.

1awesomeJ commented:
@bluca,
since the battery-check feature is near ready based on feedback from halfline.
Could you help take a look here. Two issues:
1. I don't get any journal output when i run the binary locally, I change to different log levels, I even remove the filters altogether, I keep getting "No entries in the journal"
2. It fails CI -quite woefully.

Thank you sir.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+}
+
+static int acquire_first_emergency_log_message(char **ret) {
+        sd_journal *j = NULL;
```
use `_cleanup_(sd_journal_closep)` here and you can drop the `goto clean_journal` stuff

1awesomeJ replied:
> use `_cleanup_(sd_journal_closep)` here and you can drop the `goto clean_journal` stuff

Awesome.
Thank you sir.

1awesomeJ commented:
First,
Thank you for coming back.
It's awesome ☺️☺️☺️

1awesomeJ commented:
> > 1. I don't get any journal output when i run the binary locally, I change to different log levels, I even remove the filters altogether, I keep getting "No entries in the journal"
> 
> hmm, is it possible you invoke this without privileges? i.e. as your own user? doe sit work if you run it as root?

I never tried running as root.
I'll try that. I've just been running it as "josh"

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+                goto clean_journal;
+        }
+
+        r = sd_journal_seek_head(j);
```
the head is the oldest message. we want to show the newest message. hence use seek_tail()

1awesomeJ replied:
> the head is the oldest message. we want to show the newest message. hence use seek_tail()

I thought we were going with @bluca 's suggestion of the oldest message likely being the first thing to go wrong.

You suggested newest though, but later said we should use @bluca 's suggestion

poettering replied:
ah, right, ignore me then ;-)

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+                goto clean_journal;
+        }
+
+        r = sd_journal_next(j);
```
and then _previous()

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+                log_warning_errno(r, "Failed to read next journal entry: %m");
+                goto clean_journal;
+        }
+
```
drop this empty line, as we try to keep function calls and their error handling together

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++) {
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+        }
```
maybe turn the ioctl and the for loop into a function of its own `int find_next_free_vt(int fd)`

poettering replied:
as @halfline suggested, please rework this to use VT_OPENQRY

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+
+        fd = open_terminal(tty, O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return -errno;
```
open_terminal() returns the error code in its return value, not in errno

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+
+        fd = open_terminal("/dev/tty1", O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return -errno;
```
open_terminal returns its error number in its return value, not errno

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+
+        r = ioctl(fd, VT_GETSTATE, &terminal_status);
+        if (r < 0)
+                return -errno;
```
i'd shorten this to `if (ioctl(fd, …) < 0)`

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,209 @@
+
+        r = ioctl(fd, VT_ACTIVATE, i + 1);
+        if (r < 0)
+                return -errno;
```
as above, shorten to:
```c
if (ioctl(fd, …) < 0)
        return -errno;
```

1awesomeJ commented:
> > Are we taking the oldest or newest message.
> > You earlier said we should stick with @bluca 's suggestion to fetch the oldest sir.
> 
> yeah, stick to @bluca's idea

Alright sir.

1awesomeJ commented:
@halfline 
Thank you for coming here.

I feel more confident now that I know you'll also spare some time on this.

Thank you so much sir.

halfline commented:
> @halfline Thank you for coming here.
> 
> I feel more confident now that I know you'll also spare some time on this.
> 
> Thank you so much sir.

you're welcome but  [it](https://github.com/systemd/systemd/pull/28077#discussion_r1245567265) was just a drive-by comment, i'm probably not going to be sticking around on this PR

1awesomeJ commented:
> > @halfline Thank you for coming here.
> > I feel more confident now that I know you'll also spare some time on this.
> > Thank you so much sir.
> 
> you're welcome but [it](https://github.com/systemd/systemd/pull/28077#discussion_r1245567265) was just a drive-by comment, i'm probably not going to be sticking around on this PR

Ohh...
But whenever you have some free time sir.

1awesomeJ commented:
I still don't get any output sir.
this is what I did:
```
[joshua@fedora systemd]$ ./build/systemd-bsod
No entries in the journal
Failed to acquire first emergency log message: No such file or directory
[joshua@fedora systemd]$ journalctl -b -p emerg _UID=0 -n1
-- No entries --
[joshua@fedora systemd]$ sudo logger -p emerg test-emergency-message
[sudo] password for joshua: 
[joshua@fedora systemd]$ sudo journalctl -b -p emerg _UID=0 -n1
Jun 29 10:40:45 fedora root[11442]: test-emergency-message
[joshua@fedora systemd]$ sudo ./build/systemd-bsod
No entries in the journal
Failed to acquire first emergency log message: No such file or directory
[joshua@fedora systemd]$ 
```

1awesomeJ commented:
I tried switching to ```sd_journal_seek_tail()``` and ```sd_journal_previous()```.
It's the same situation.

1awesomeJ commented:
> > you're welcome but [it](https://github.com/systemd/systemd/pull/28077#discussion_r1245567265) was just a drive-by comment, i'm probably not going to be sticking around on this PR
> 
> @halfline thanks so much for all the help + reviews on the other PR, very much appreciated!

Yes sir.
He spent much time on it.

1awesomeJ commented:
Looks like the issue is with the formatting of the _BOOT_ID filter.
I downloaded the ```systemd-devel``` package and tried to build another tool outside my systemd repo.

It displayed the emergency message successfully.
```
[joshua@fedora ~]$ sudo journalctl -b -p emerg _UID=0 -n1
Jun 29 10:40:45 fedora root[11442]: test-emergency-message
[joshua@fedora ~]$ vi read_journal.c
[joshua@fedora ~]$ gcc -o read_journal read_journal.c -lsystemd
[joshua@fedora ~]$ sudo ./read_journal 
MESSAGE=test-emergency-message
```

1awesomeJ commented:
I still haven't figured why the CI fails though.

poettering commented:
i tried your current version here, and it worked fine actually.

1awesomeJ commented:
> i tried your current version here, and it worked fine actually.

Wow.

I should consider doing a dual boot on my new computer. I've been using the old one to run Fedora.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+        const void *d;
+        size_t l;
+        sd_id128_t boot_id;
+        char boot_id_filter[42];
```
this is your bug. you need 43 chars here.

encode as:

```c
char boot_id_filter[STRLEN("_BOOT_ID=") + SD_ID128_STRING_MAX];
```

to avoid any confusion.

1awesomeJ replied:
> this is your bug. you need 43 chars here.
> 
> encode as:
> 
> ```c
> char boot_id_filter[STRLEN("_BOOT_ID=") + SD_ID128_STRING_MAX];
> ```
> 
> to avoid any confusion.


Wow...
😁😁😁😁

I'm glad I'm learning from the very best.
I'll cherish the memories.

Thank you sir.

1awesomeJ commented:
Thank you for taking the time to test sir.

I still need to be able to run it though, so I see the VT open, make it turn blue,  and then move to the QR encoding.


Do the CI failures have to do with my return value possible being non-zero?

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the current boot with an emergency log level.\n\n"
```
this needs to be line broken. one of the tests in the testuite is tripping up on this, because we want that help texts are 79ch wide at max to deal with traditional terminals which used to be 80x25 wide.

1awesomeJ replied:
Ohhh....
That's interesting.

Thank you sir.




Let me fix and push.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        *ret = strndup((char *) d, l);
```
we have the rule that functions should never touch return parameters on failure, and strictly initialize all return parameters on succeed. Here though, if the strdup() you will fail, but already touched `ret`.

hence, please make the copy into temporary variable, and once it worked, and nothing can fail anymore, move it into `ret`

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+        *ret = strndup((char *) d, l);
+        if (! *ret) {
+                r = -ENOMEM;
+                return log_error_errno(r, "Failed to duplicate journal message: %m");
```
just use `return log_oom()`

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+static int display_emergency_message_fullscreen(char * message) {
+        assert(message);
+        int r;
+        char tty[20];
```
use `char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1]` to size the string properly

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to find a free VT: %m");
+
+        snprintf(tty, sizeof(tty), "/dev/tty%d", r + 1);
```
xsprintf

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+                return log_error_errno(fd, "Failed to open tty: %m");
+
+        if (ioctl(fd, VT_ACTIVATE, r + 1) < 0)
+                return -errno;
```
please log here, so that all errors are logged about

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,195 @@
+
+        r = loop_write(fd, message, strlen(message), false);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to write emergency message to terminal.: %m");
```
extra `.`

1awesomeJ commented:
Wow...
Thank you so much for this sir.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,199 @@
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = strndup((char *) d, l);
+        if (! *ret)
```
need to check `message` here, not `*ret`

1awesomeJ replied:
> need to check `message` here, not `*ret`

I forgot to edit 😂😂

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,199 @@
+        r = loop_write(fd, message, strlen(message), false);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to write emergency message to terminal: %m");
+
```
please wait for a key press here, and if you receive it, switch back to the original VT

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,199 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = strndup((char *) d, l);
```
wow, I'll get so familiar with the codebase much faster this way-Learning about these cool functions.

1awesomeJ commented:
The CI failures though. It's moved from "woeful"(24 failures) to "bad"(maybe 15).

1awesomeJ commented:
@halfline , @poettering 
About ```VT_OPENQRY``` and  ```VT_GETSTATE```
I modified our ```ioctl()``` call to not help us find an available tty, but also help know identify the currently active tty so we can switch back to it on keypress.

Could you take a look to see if the current usage justifies ```VT_GETSTATE``` or we should still stick with ```VT_OPENQRY``` and find another way to identify the currently active terminal?

Thank you sirs.

halfline commented:
> @halfline , @poettering About `VT_OPENQRY` and `VT_GETSTATE` I modified our `ioctl()` call to not help us find an available tty, but also help know identify the currently active tty so we can switch back to it on keypress.

Ah I see... maybe okay then.

If you want to be more bulletproof you could use both. First call `VT_OPENQRY` to get a vt number, then use that vt number to find the associated tty. Open that tty, so your file descriptor "reserves" the VT. Then use `VT_GETSTATE` *after opening it* to make sure the VT didn't get taken by someone else after your `VT_OPENQRY` call and to find the current active VT. Then use `VT_SETACTIVATE` (instead of `VT_ACTIVATE`) to jump to the tty associated with the tty file descriptor.

maybe overkill when there's not likely to be more than one thing mucking with the VTs at the same time, I guess. not sure.  

I do have some vague notion that it's unkosher to the kernel for userspace to call `VT_ACTIVATE` without pairing it with a `VT_WAITACTIVE` call afterward but I don't remember where I got that notion from, so I might be wrong. But if so, switching to `VT_SETACTIVATE` would probably sidestep whatever that problem is (if it is a problem and i'm not crazy)

could just keep things simple and revisit if there's a problem in the future i suppose.

1awesomeJ commented:
> > @halfline , @poettering About `VT_OPENQRY` and `VT_GETSTATE` I modified our `ioctl()` call to not help us find an available tty, but also help know identify the currently active tty so we can switch back to it on keypress.
> 
> Ah I see... maybe okay then.
> 
> If you want to be more bulletproof you could use both. First call `VT_OPENQRY` to get a vt number, then use that vt number to find the associated tty. Open that tty, so your file descriptor "reserves" the VT. Then use `VT_GETSTATE` _after opening it_ to make sure the VT didn't get taken by someone else after your `VT_OPENQRY` call and to find the current active VT. Then use `VT_SETACTIVATE` (instead of `VT_ACTIVATE`) to jump to the tty associated with the tty file descriptor.
> 
> maybe overkill when there's not likely to be more than one thing mucking with the VTs at the same time, I guess. not sure.
> 
> I do have some vague notion that it's unkosher to the kernel for userspace to call `VT_ACTIVATE` without pairing it with a `VT_WAITACTIVE` call afterward but I don't remember where I got that notion from, so I might be wrong. But if so, switching to `VT_SETACTIVATE` would probably sidestep whatever that problem is (if it is a problem and i'm not crazy)
> 
> could just keep things simple and revisit if there's a problem in the future i suppose.

Wow...
Thank you sir.
The notion is certainly from a good place 😁.

I'll look up VT_SETACTIVATE.
From what you've stated sir, I don't see any harm/cost switching to it.

I'll look it up and implement all.

Thanks again sir.

halfline commented:
> I'll look up VT_SETACTIVATE. From what you've stated sir, I don't see any harm/cost switching to it.

Well just that it's more added complexity for a rare corner case where two things are trying to get a new vt at nearly the same time.  Add truthfully, if an X server or something is trying to allocate a VT at the same time as this bsod screen, the user experience is going to take a hit whether or not it and this end up on the same vt or not. if they each correctly got their own VT, there would still be a race for which process wins at setting the active VT; the bsod screen may or may not show regardless.

So the extra complexity won't even fully resolve the contention problem. and i'm guessing the contention problem is unlikely to manifest in practice anyway.

1awesomeJ commented:
> > I'll look up VT_SETACTIVATE. From what you've stated sir, I don't see any harm/cost switching to it.
> 
> Well just that it's more added complexity for a rare corner case where two things are trying to get a new vt at nearly the same time. Add truthfully, if an X server or something is trying to allocate a VT at the same time as this bsod screen, the user experience is going to take a hit whether or not it and this end up on the same vt or not. if they each correctly got their own VT, there would still be a race for which process wins at setting the active VT; the bsod screen may or may not show regardless.
> 
> So the extra complexity won't even fully resolve the contention problem. and i'm guessing the contention problem is unlikely to manifest in practice anyway.

Perhaps we should wait for when users raise any occurrence of this rare contention problem in the future, then we'll figure out a better fix for it?
Their description of the problem will probably give us more context then?

Please check this PR again soon to help guide on our implementation of the QR code.
Alongside displaying the emergency message on the blue screen, we also want to display a QR code that users can scan with their phones to figure out what caused their boot failures.

Thank you so much for your time sir.

1awesomeJ commented:
We'll be sure to call your attention sir.

halfline commented:
> If you want to be more bulletproof you could use both. First call `VT_OPENQRY` to get a vt number, then use that vt number to find the associated tty. Open that tty, so your file descriptor "reserves" the VT. Then use `VT_GETSTATE` _after opening it_ to make sure the VT didn't get taken by someone else after your `VT_OPENQRY` call and to find the current active VT.

Thinking about this more it's wrong headed... the `VT_GETSTATE` will always show it in use in this case because (because we've reserved it!)... so now that i've said it outloud and thought about it, I don't think it's a good idea.

I think your approach is right and we shouldn't have `VT_OPENQRY` in the picture at all.

1awesomeJ commented:
@poettering, Good afternoon sir.
I still couldn't figure out why I don't get lg outputs.
When I directly run the executable from the ninja build.
However, when I write the code in just a
regular file in my home directory, and compile with "-lsystemd", the executable I get fetches the emergency message accurately.
So I plan using that as an approximation going forward, so I can increase my pace on the project.
The only catch is I lose access t all our cool custom functions, especially the error logging functions.
I hope to find a fix soon though.

Here's what I approximated the current "acquire_first_emergency_log_message()" function to:

```
#include <systemd/sd-journal.h>
#include <stdio.h>
#include <stdlib.h>

static int acquire_first_emergency_log_message(char **ret) {
    sd_journal *j = NULL;
    const void *d;
    size_t l;
    sd_id128_t boot_id;
    char boot_id_filter[44];
    int r;

    r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
    if (r < 0) {
        printf("Failed to open journal\n");
        return r;
    }

    r = sd_id128_get_boot(&boot_id);
    if (r < 0) {
        printf("Failed to get boot ID\n");
        sd_journal_close(j);
        return r;
    }

    snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));

    r = sd_journal_add_match(j, boot_id_filter, 0);
    if (r < 0) {
        printf("Failed to add boot ID filter\n");
        sd_journal_close(j);
        return r;
    }

    r = sd_journal_add_match(j, "_UID=0", 0);
    if (r < 0) {
        printf("Failed to add User ID filter\n");
        sd_journal_close(j);
        return r;
    }

    r = sd_journal_add_match(j, "PRIORITY=0", 0);
    if (r < 0) {
        printf("Failed to add Emergency filter\n");
        sd_journal_close(j);
        return r;
    }

    r = sd_journal_seek_head(j);
    if (r < 0) {
        printf("Failed to seek to start of journal\n");
        sd_journal_close(j);
        return r;
    }

    r = sd_journal_next(j);
    if (r < 0) {
        printf("Failed to read next journal entry\n");
        sd_journal_close(j);
        return r;
    } else if (r == 0) {
        printf("No entries in the journal\n");
        sd_journal_close(j);
        return -1;
    }

    r = sd_journal_get_data(j, "MESSAGE", &d, &l);
    if (r < 0) {
        printf("Failed to read journal message\n");
        sd_journal_close(j);
        return r;
    }

    *ret = strndup(d, l);
    if (!*ret) {
        printf("Out of memory\n");
        sd_journal_close(j);
        return -1;
    }

    sd_journal_close(j);
    return 0;
}

int main(void) {
    char *message = NULL;
    int r;

    r = acquire_first_emergency_log_message(&message);
    if (r == 0) {
        printf("THIS IS THE LOG MESSAGE: %s\n", message);
    }

    free(message);
    return r;
}
```

1awesomeJ commented:
This fetches the message only when I compile it from outside my local systemd repository.
Now I'll move t testing the tty display function.

1awesomeJ commented:
@poettering @bluca ,

Good morning sirs.

The error message does display full screen on the tty now.

But a number of issues:

1. How can we position it better? It sorts of just sticks to the top left corner.
Can we alter that?

2. How do we make the tty turn blue, it's currently black.

3. I'm thinking of concatenating an instruction to the message that says "press anybkey to exit"
4. I approximated our read_one_char(stream); with fgetc(stream);
I'm not able to exit on keypress yet 
![IMG_20230704_101544_906](https://github.com/systemd/systemd/assets/109356020/60b14ca9-0920-41fa-95ea-02e5d60155e5)

1awesomeJ commented:
The issues with the key press was that I had to also hit "\n".
Something we were able to turn off with our read_one_char().

Let me now see if I can attempt the blue screen while I wait!!!

1awesomeJ commented:
![IMG_20230704_130735_521](https://github.com/systemd/systemd/assets/109356020/11c538e3-7aab-4baa-a6da-e76ea812ec77)

1awesomeJ commented:
I got the terminal to turn blue although via a crude method of writing ANSI_BLUE_BACKGROUND in loops 80 by 40 times. I tried 80 by 24 it covered only a quarter of my screen.

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,217 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = (char *)memdup_suffix0(d, l);
```
no need to cast, void* implicitly can be converted to any other pointer type

poettering replied:
you probably want to cut off the `MESSAGE=` prefix here:

```c
message = memdup_suffix0((const char*) d + STRLEN("MESSAGE="), l - STRLEN("MESSAGE="));
```

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,217 @@
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++) {
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+        }
```
you can drop the `{}` btw

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,217 @@
+        return r;
+}
+
+static int find_next_free_vt(int fd, int *free_vt, int *original_vt) {
```
we generally have this rule that return parameters should be called `ret` or `ret_xyz`. Hence please change `free_vt` and `original_vt` to `ret_free_vt`, `ret_original_vt`

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,217 @@
+        if (ioctl(fd, VT_ACTIVATE, free_vt + 1) < 0)
+                return log_error_errno(errno, "Failed to activate tty: %m");
+
+        r = loop_write(fd, ANSI_HOME_CLEAR, strlen(ANSI_HOME_CLEAR), false);
```
hm, if you output ANSI_BACKGROUND_BLUE before you issue ANSI_HOME_CLEAR, doesn't this make the whole screen background blue?

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,217 @@
+
+        r = read_one_char(stream, NULL, USEC_INFINITY, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failled to read character: %m");
```
typo: failled → failed

poettering commented:
> I got the terminal to turn blue although via a crude method of writing ANSI_BLUE_BACKGROUND in loops 80 by 40 times. I tried 80 by 24 it covered only a quarter of my screen.

BTW, the ANSI sequences we can use on the VT are documented (incompletely) here:

https://man7.org/linux/man-pages/man4/console_codes.4.html

i.e. our ANSI_HIGHLIGHT macros generate the sequences listed under "ECMA-48 CSI sequences" mostly.

poettering commented:
> * How can we position it better? It sorts of just sticks to the top left corner.
>   Can we alter that?
> 
> * How do we make the tty turn blue, it's currently black.
> 
> * I'm thinking of concatenating an instruction to the message that says "press anybkey to exit"
> 
> * I approximated our read_one_char(stream); with fgetc(stream);
>   I'm not able to exit on keypress yet

You can position the cursor via ANSI sequences, see link I just posted.

In order to position the text in the middle of the screen you can use TIOCGWINSZ ioctl.

1awesomeJ commented:
Yay!!!!!

Thank you so much sir.

Always awesome to get your reviews.
Thank you so much sir.

1awesomeJ commented:
![IMG_20230706_134159_641](https://github.com/systemd/systemd/assets/109356020/ccaec9d7-b4d0-4bfe-a7ae-7c24f441a568)

1awesomeJ commented:
@poettering , Thanks again for the constant guidance. This mentorship has been the best thing to happen to me this year. "Thank you" to you all.

I've implemented all the requested changes and I've pushed the patch. 
I also tested using my approximate method which mostly is same code but with logging done with printf.

Since we don't know how long a message may be, I didn't implement horizontal positioning.
However, I added a util, with which we can move the cursor vertically.

If you know how much I struggled earlier to get blue screen, then you came and told me how to do it in just one line of code,
Thank you so much sir!!!

1awesomeJ commented:
I'll now study for the QR code implementation while I await your review and guidance.
Thank you sir.

bluca commented on man/systemd-bsod.xml:
```diff
@@ -0,0 +1,62 @@
+  "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
+<!-- SPDX-License-Identifier: LGPL-2.1-or-later -->
+
+<refentry id="systemd-ac-power" xmlns:xi="http://www.w3.org/2001/XInclude">
```
id needs to be updated

1awesomeJ replied:
Wow... 

Thank you sir.

1awesomeJ commented:
@bluca 
It almost seemed like you said,
"I'll just leave those two to their business, they seem to get along just fine"

Thanks for showing up sir.

bluca commented:
> I'll now study for the QR code implementation while I await your review and guidance. Thank you sir.

For QR code we already have support for libqrencode, so check out src/shared/qrcode-util.h

1awesomeJ commented:
> > I'll now study for the QR code implementation while I await your review and guidance. Thank you sir.
> 
> For QR code we already have support for libqrencode, so check out src/shared/qrcode-util.h

And we're back!!! 
😁😁😁

Thank you sir.

1awesomeJ commented:
@bluca, @poettering 
Good morning sirs,
1. I added the QR code, I've pushed, but I couldn't test locally, it says "Operation not supported".
2. About testing, the main issue I had was I could never fetch a message from the journal while running the executable in build/.
However, I fetched successfully, using the same code compiled with "systemd-devel" outside of my local repo. So I decided to test the rest of the code using a 
hard coded message, I was able to find an issue with the call to ```read_one_char()```, I had been passing NULL as it's second param, rather than the 
address of a buffer where it could store the character read. so ```assert()``` was failing on the NULL.
3. There was another assert failure which I'm yet t figure out. This is the exact message "Assertion 'fclose_nointr(f) != -EBADF' failed at src/basic/fd-util.c:126, function safe_fclose(). Aborting.
Aborted
"
4. The battery service from the first PR, I kept it running(out of excitement). It's still enabled on my device. So on my last boot, I saw the updated A.C. message -for the first time. Maybe my plymouth eventually decided to open its mail today.

5. Will a work around for this QR code involve a "force-print-QR", like we did with the glyph?

Thank you sirs.

bluca commented:
Make sure you have libqrencode-devel (or however it is called on the distro you are building on) installed, and after that do `meson configure build/ -Dqrencode=true` and rebuild

1awesomeJ commented:
@bluca , sorry I didn't give any updates over the weekend. I was planning a countryside
![IMG_20230710_135458_238](https://github.com/systemd/systemd/assets/109356020/b1093ca3-63dd-4ed3-a50c-220392deb83e)
 trip.

I implemented your guidance, and the QR code shows clearly now.

It's super cool!!!!

bluca commented:
Very nice

1awesomeJ commented:
It was also accurately decoded when ii scanned with my phone.

Thank you for saving me hours/days of hard search for a fix

1awesomeJ commented:
Good morning sirs.
Here's my update:

1. About the URL, I hard coded the freedesktop wiki page, it works well, the QR scanner loads the web page automatically. What I await now is how to implement a specific URL for each emergency message we're displaying.

2. About the positioning, I can't seem to move the print horizontally.  I modified the
cursor function to allow horizontal and vertical positioning, I see the cursor blink at any position I set, but the QR code only moves vertically, it won't shift from the left edge of the screen at all. Makes me think I'm missing something.
![IMG_20230712_032026_662](https://github.com/systemd/systemd/assets/109356020/8ac9c1d2-3a8a-4d56-bac1-0b2427cd01ff)
![IMG_20230712_032552_301](https://github.com/systemd/systemd/assets/109356020/78429916-fc39-4609-a3d9-cfe718148e69)

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+
+        if (!message) {
+                log_debug("No entries in the journal");
+                return r;
```
This should return 0

poettering commented on src/basic/terminal-util.c:
```diff
@@ -1476,3 +1476,15 @@ void get_log_colors(int priority, const char **on, const char **off, const char
         }
 }
+
+int set_terminal_cursor_position(int fd, int row, int column) {
```
hmm, i'd really make the row/column parameters `unsigned`. They cannot possibly be negative after all...

poettering commented on src/basic/terminal-util.c:
```diff
@@ -1476,3 +1476,15 @@ void get_log_colors(int priority, const char **on, const char **off, const char
+        char cursor_position[STRLEN("\x1B[") + DECIMAL_STR_MAX(int) * 2 + STRLEN(";H") + 1];
+        xsprintf(cursor_position, "\x1B[%d;%dH", row, column);
+
+        r = write(fd, cursor_position, strlen(cursor_position));
```
better use loop_write()

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to change terminal color to blue, ignoring: %m");
+
+        r = loop_write(fd, ANSI_HOME_CLEAR, strlen(ANSI_HOME_CLEAR), false);
```
maybe merge these two loop_write() expressions, by concatenating the strings? i.e.:

```c
r = loop_write(fd, ASNI_BACKGROUND_BLUE ANSI_HOME_CLEAR, strlen(ANSI_BACKGROUND_BLUE ANSI_HOME_CLEAR), /* poll= */ false);
```

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+                log_warning_errno(r, "Failed to clear terminal, ignoring: %m");
+
+        qr_code_start_row = (int) (w.ws_row * 0.75);
+        qr_code_start_column = (int) (w.ws_col * 0.75);
```
if TIOCGWINSZ failed you'll access uninitialized memory here. maybe just initialize `w` to 80x25 in case of failure, i.e. the traditional VGA text console size from the 1990s.

i usually would avoid floating point calculations if we don't need to. To take 3/4th of something just write it out as integer calculations:

`w.ws_row * 3U / 4U`

poettering commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        if (out == NULL)
+                return log_error_errno(errno, "Failed to open output file: %m");
+
+        r = print_qrcode(out, "Scan the code for help", "https://www.freedesktop.org/wiki/");
```
print_qrcode() as it is right now, can only write stuff fully left aligned on screen. 

i guess you have to extend the function to optionally take a cursor position x/y as arguments. And if specified issue set_terminal_cursor_position() before each line.

poettering replied:
and the string in the QR code should really be the error string, and maybe some more info, dependig how much space we have.

poettering reviewed (changes requested):
this looks excellent already, but I think this needs some alignment love. Maybe put the text version in the top left (maybe with some space to the left and to the top), and the QR code bottom right (also with some space to the right and the bottom).

1awesomeJ commented:
Wow.....
Sorry I've not been very much online today.
I'm still on my countryside trip.

I'll certainly work on all the guidance beginning tonight.


My target too is bottom right for the QR code.

Thank you so much sirs.

1awesomeJ commented:
I'm so privileged.
Going through the reviews again, now that I'm more relaxed.
I'm super excited and grateful.

Thank you so much sirs.

1awesomeJ commented:
@poettering , @bluca , Good morning sirs.
I pushed the requested changes.
The following are outstanding though:
1. I need to modify ```print_qrcode() ``` to make it lose its current left-screen-side affinity.
2. We need to make our QR code more helpful, my thought is if the user can already read the message on their screen, our code should have something more helpful than just the emergency message which they can already read. The experience may actually be disappointing for some users who see a cool QR code, only to find that it contains just a string they've already read.

What I would like is if scanning the code could sort of open a web page or man page for the user, that has a 70%+ chance of helping them figure out the problem with the boot.

3. I need to begin working on Poettering's guidance of linking this tool up with journald.

4. There's also still the issue of the code not displaying by default, I had to download the qrencode devel package, and reconfigure meson to see it. How do we fix those for all users?

Thank you sirs!!!

bluca commented:
Only the first one needs to be fixed in this PR, the rest can be done later. The package installation is an integration issue, not our problem to solve, we just document the dependency and let the image builders sort it out

bluca commented:
```
977s ../src/journal/bsod.c: In function ‘main’:
977s ../src/journal/bsod.c:169:13: error: ‘original_vt’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
977s   169 |         if (ioctl(fd, VT_ACTIVATE, original_vt) < 0)
977s       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
977s ../src/journal/bsod.c:109:25: note: ‘original_vt’ was declared here
977s   109 |         int r, free_vt, original_vt/*, qr_code_start_row = 1, qr_code_start_column = 1*/;
977s       |                         ^~~~~~~~~~~
977s In file included from ../src/basic/log.h:13,
977s                  from ../src/basic/macro.h:446,
977s                  from ../src/basic/alloc-util.h:10,
977s                  from ../src/journal/bsod.c:9:
977s ../src/basic/stdio-util.h:26:27: error: ‘free_vt’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
977s    26 |         assert_message_se(snprintf_ok(buf, ELEMENTSOF(buf), fmt, ##__VA_ARGS__), "xsprintf: " #buf "[] must be big enough")
977s       |                           ^~~~~~~~~~~
977s ../src/journal/bsod.c:109:16: note: ‘free_vt’ was declared here
977s   109 |         int r, free_vt, original_vt/*, qr_code_start_row = 1, qr_code_start_column = 1*/;
977s       |                ^~~~~~~
```

1awesomeJ commented:
> Only the first one needs to be fixed in this PR, the rest can be done later. The package installation is an integration issue, not our problem to solve, we just document the dependency and let the image builders sort it out

Ohh...
Okay sir.

1awesomeJ commented:
> ```
> 977s ../src/journal/bsod.c: In function ‘main’:
> 977s ../src/journal/bsod.c:169:13: error: ‘original_vt’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
> 977s   169 |         if (ioctl(fd, VT_ACTIVATE, original_vt) < 0)
> 977s       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 977s ../src/journal/bsod.c:109:25: note: ‘original_vt’ was declared here
> 977s   109 |         int r, free_vt, original_vt/*, qr_code_start_row = 1, qr_code_start_column = 1*/;
> 977s       |                         ^~~~~~~~~~~
> 977s In file included from ../src/basic/log.h:13,
> 977s                  from ../src/basic/macro.h:446,
> 977s                  from ../src/basic/alloc-util.h:10,
> 977s                  from ../src/journal/bsod.c:9:
> 977s ../src/basic/stdio-util.h:26:27: error: ‘free_vt’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
> 977s    26 |         assert_message_se(snprintf_ok(buf, ELEMENTSOF(buf), fmt, ##__VA_ARGS__), "xsprintf: " #buf "[] must be big enough")
> 977s       |                           ^~~~~~~~~~~
> 977s ../src/journal/bsod.c:109:16: note: ‘free_vt’ was declared here
> 977s   109 |         int r, free_vt, original_vt/*, qr_code_start_row = 1, qr_code_start_column = 1*/;
> 977s       |                ^~~~~~~
> ```

I didn't notice this when I built. I'll fix sir.

1awesomeJ commented:
Excerpt from v254's release notes as seen from @yuwata's cool tweet:
```
    * systemd-cryptenroll and other tools that take TPM2 PCR parameters now
      understand textual identifiers for these PCRs.
 ```

Looks like I had a small part in this. Super awesome.
I'll wrap up this trip in about 9 days, pick up pace, and hope to have even more part in v255.
 Thank you for the privilege!

1awesomeJ commented:
Not there yet, but I got it to move:

![IMG_20230716_130740_769](https://github.com/systemd/systemd/assets/109356020/f9d75226-afdc-41b4-85a9-16309f52ff92)

1awesomeJ commented:
@bluca , @poettering 
Good evening sirs.
This is what I came up with:

![Uploading IMG_20230716_180836_805.jpg…]()

1awesomeJ commented:
1. Somehow, a blank row still appears between the top border and the actual code.
2. The string that says "scan this code" isn't positioned directly above the print as I expect.


So let me push the patch, so you can take a look.
Plus I modified up to three functions, perhaps I can be more efficient.

Thank you sirs

bluca commented:
```
../src/journal/bsod.c: In function ‘display_emergency_message_fullscreen’:05:22
../src/journal/bsod.c:158:13: error: implicit declaration of function ‘print_positioned_qrcode’ [-Werror=implicit-function-declaration]05:22
  158 |         r = print_positioned_qrcode(out, "Scan the code for help", (const char *) message, qr_code_start_row, qr_code_start_column);05:22
      |             ^~~~~~~~~~~~~~~~~~~~~~~
```

this needs to be made conditional on the qr library being available. Simply skip building the whole thing in Meson if not available, no need for a fallback, as it's an optional component

1awesomeJ commented:
> ```
> ../src/journal/bsod.c: In function ‘display_emergency_message_fullscreen’:05:22
> ../src/journal/bsod.c:158:13: error: implicit declaration of function ‘print_positioned_qrcode’ [-Werror=implicit-function-declaration]05:22
>   158 |         r = print_positioned_qrcode(out, "Scan the code for help", (const char *) message, qr_code_start_row, qr_code_start_column);05:22
>       |             ^~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> this needs to be made conditional on the qr library being available. Simply skip building the whole thing in Meson if not available, no need for a fallback, as it's an optional component

Ohh...okay sir.
Happy new week sir.

bluca commented on src/basic/terminal-util.c:
```diff
@@ -1476,3 +1476,16 @@ void get_log_colors(int priority, const char **on, const char **off, const char
+int set_terminal_cursor_position(int fd, unsigned int row, unsigned int column) {
+        int r;
+        char cursor_position[STRLEN("\x1B[") + DECIMAL_STR_MAX(int) * 2 + STRLEN(";H") + 1];
+        xsprintf(cursor_position, "\x1B[%u;%uH", row, column);
```
add empty line above

bluca replied:
assert(fd >= 0)

bluca commented on src/basic/terminal-util.c:
```diff
@@ -1476,3 +1476,16 @@ void get_log_colors(int priority, const char **on, const char **off, const char
+        xsprintf(cursor_position, "\x1B[%u;%uH", row, column);
+
+        r = loop_write(fd, cursor_position, strlen(cursor_position), /* do_poll = */false);
+
```
remove this empty line

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+static int acquire_first_emergency_log_message(char **ret) {
+        _cleanup_(sd_journal_closep) sd_journal *j = NULL;
+        _cleanup_free_ char *message = NULL;
+
```
remove this empty line

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        sd_id128_t boot_id;
+        char boot_id_filter[STRLEN("_BOOT_ID=") + SD_ID128_STRING_MAX];
+        int r;
+
```
assert(ret)

1awesomeJ replied:
> assert(ret)

This would mean, we have to always initialize any pointer whose address we pass to this function?

bluca replied:
no, only the return storage

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to read next journal entry: %m");
+        if (r == 0) {
+                log_debug("No entries in the journal");
```
"No emergency level entries in the journal"

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+                return log_error_errno(r, "Failed to read journal message: %m");
+
+        message = memdup_suffix0((const char*)d + STRLEN("MESSAGE="), l - STRLEN("MESSAGE="));
+        if (! message)
```
remove space after `!`

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+
+        *ret = TAKE_PTR(message);
+
+        return r;
```
return 0

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+
+static int find_next_free_vt(int fd, int *ret_free_vt, int *ret_original_vt) {
+        assert(fd);
+
```
assert(ret_free_vt) and ret_original_vt, and move below declarations

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+
+        *ret_free_vt = i;
```
what happens if no free vt was found?

1awesomeJ commented:
Alright sir.
Let me make these quick fixes right away.
I've been trying to figure out varlink and our JSON library.
There seems to be a lot of functions and data structure to learn there.

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        return 0;
+}
+
+static int display_emergency_message_fullscreen(char * message) {
```
remove space after `*`

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+}
+
+static int display_emergency_message_fullscreen(char * message) {
+        assert(message);
```
move assert after declarations

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,244 @@
+        if (r < 0)
+                return log_warning_errno(r, "Failed to write emergency message to terminal: %m");
+
+        out = fdopen(fd, "w");
```
fdopen will take possession of the fd, use fdopen_independent()

bluca replied:
open with `r+` so you don't need a separate file stream for reading and writing

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -36,31 +36,45 @@ int dlopen_qrencode(void) {
+        if (fd < 0)
+                return (void)log_warning_errno(fd, "unable to get file descriptor from the file stream: %m");
+
+        set_terminal_cursor_position(fd, row, column);
```
these should be done only if row and column are not 0

1awesomeJ commented:
@bluca,
I've pushed the requested changes. All of them but:
1. ```assert(ret)``` in ```static int acquire_first_emergency_log_message(char **ret)```: we would only be able to call with the address of a non-NULL string? the pointer would have to be initialized to a non-NULL string?
2. ```assert(ret_free_vt)``` and ```ret_original_vt``` in ```static int find_next_free_vt(int fd, int *ret_free_vt, int *ret_original_vt)```: same reason as 1. Since we're not using their values at all, but rather just using them to fetch data, do we have to assert on them?
3. The optional meson build when the environment doesn't support qr code functionality: I haven't settled down to understand how to implement it, as well as the final fix to the ```print_positioned_qr_code```


I think all the other ones have been sorted. I'm sorry I still make these blank line/white space mistakes.

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,249 @@
+                        break;
+
+        if (i >= sizeof(terminal_status.v_state) * 8)
+                return log_error_errno(-1, "No free VT found: %m");
```
`log_error_errno(SYNTHETIC_ERRNO(ENOTTY)`

1awesomeJ replied:
Thank you so much sir!

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -36,31 +36,47 @@ int dlopen_qrencode(void) {
                         fputs(UNICODE_FULL_BLOCK, output);
 
                 fputs(ANSI_NORMAL "\n", output);
+                set_terminal_cursor_position(fd, row + 1, column);
```
all of these new operations need to happen only if row or column are set, here and below

1awesomeJ replied:
> all of these new operations need to happen only if row or column are set, here and below

I think I'll split up to two functions: The normal ```print_qrcode``` that we call when no position is to be set, and the ```print_positioned_qrcode``` that gets called when the QR code's position is specified.

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -61,6 +81,7 @@ static void write_qrcode(FILE *output, QRcode *qr) {
                 const uint8_t *row2 = row1 + qr->width;
 
                 fputs(ANSI_WHITE_ON_BLACK, output);
+
```
spurious new line

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -48,6 +48,26 @@ static void print_border(FILE *output, unsigned width) {
+        if (fd < 0)
+                return (void)log_warning_errno(fd, "unable to get file descriptor from the file stream: %m");
+
+        if (row > 0 || column > 0)
```
this check is no longer necessary now

1awesomeJ replied:
> this check is no longer necessary now

Ohh...
Let me fix

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,249 @@
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the\n"
+               " current boot with an emergency log level and displays it"
```
missing newline character

bluca replied:
and extra space at the beginning

1awesomeJ replied:
> and extra space at the beginning

Oh...I just broke the string to two lines quite hastily.

1awesomeJ commented:
@bluca,
Good morning sir.
Thank you so so so much for spending time with me yesterday.
I'm sorry I dragged you through a lot of petty style issues.


I'll be going back to Lagos on Sunday.
Maybe I'll take a day or two to rest and get my mind reset to studying mode,
By next week, I'll have a better pace, and less cases of terrible coding style.

Thanks again.

I was gonna ask again about streaming the "All Systems Go" conference,
but I saw it's not happening till mid September?

If I can't watch you live, I'll just watch some recorded videos from this or other events.

Thank you for being great mentors.

bluca commented:
Still needs to be made optional depending on the presence of libqrencode

1awesomeJ commented:
> Still needs to be made optional depending on the presence of libqrencode

Certainly sir. The print_positioned_qrcode() too still has an alignment issue. 

I'll work on those two sir

1awesomeJ commented:
Good morning I got back to Lagos around midnight yesterday (34 hours ago), I ate and slept yesterday, I have now resumed studying.

bluca commented on src/shared/qrcode-util.h:
```diff
@@ -7,9 +7,14 @@
 #if HAVE_QRENCODE
 int dlopen_qrencode(void);
 
+int print_positioned_qrcode(FILE *out, const char *header, const char *string, unsigned int row, unsigned int column);
 int print_qrcode(FILE *out, const char *header, const char *string);
 #else
 static inline int print_qrcode(FILE *out, const char *header, const char *string) {
         return -EOPNOTSUPP;
 }
+
+static inline int print_positioned_qrcode(FILE *out, const char *header, const char *string) {
+        return -EOPNOTSUPP;
+}
 #endif
```
also, but not only, building the bsod binary itself needs to be optional depending on whether the library is available by checking `HAVE_QRENCODE` in meson.build, and not be built at all if it's not set

1awesomeJ replied:
> also, but not only, building the bsod binary itself needs to be optional depending on whether the library is available by checking `HAVE_QRENCODE` in meson.build, and not be built at all if it's not set

ohh... makes sense. I found people will be paying hundreds of Euros to listen to you guys for just 2 days in September, and I am even getting paid while having this direct access for months now. I am excited and grateful. Thank you sirs.

1awesomeJ commented on meson.build:
```diff
@@ -4004,6 +4004,20 @@ public_programs += executable(
         install_rpath : rootpkglibdir,
         install : true)
 
+if conf.get('HAVE_QRENCODE') == 1
+        public_programs += executable(
+                'systemd-bsod',
+                'src/journal/bsod.c',
+                include_directories : includes,
+                link_with : [libshared],
+                dependencies : [userspace,
+                                versiondep,
+                                libqrencode],
+                install_rpath : rootpkglibdir,
+                install_dir : rootlibexecdir,
+                install : true)
+endif
```
@bluca , is this it sir?

bluca replied:
yes that's fine

1awesomeJ replied:
> yes that's fine

Okay sir.

I'll now work on the alignment fix.

1awesomeJ commented:
After pretty much lots of trials and errors, I now have the qr code at the bottom left.

I'll look deeper to see how to clean it up ,and standardize.
![IMG_20230725_162313_338](https://github.com/systemd/systemd/assets/109356020/26b20f7c-80e1-43c3-842f-b8d7dcd99f4c)

1awesomeJ commented:
This is even further to the right.
![IMG_20230725_162800_826](https://github.com/systemd/systemd/assets/109356020/184a6685-4f73-4cf1-bcb9-78f6a07582e9)

1awesomeJ commented:
@bluca, Good morning sir. This current patch prints the QR code at the bottom right. But it has about three issues, I'll highlight them.

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,249 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to clear terminal, ignoring: %m");
+
+        qr_code_start_row = w.ws_row * 3U / 5U;
+        qr_code_start_column = w.ws_col * 3U / 4U;
```
1. 60% height 75% width was gotten as the optimal position for my terminal window. lower numbers wouldn't be bottom-left enough, higher numbers would scatter the QR code drawing. I'd like to put some sort of fallback in place so that if 60%/75% doesn't work for a terminal, the fallback value is used and the QR code is saved.

bluca replied:
that's probably good enough for now

1awesomeJ commented on src/shared/qrcode-util.c:
```diff
@@ -89,6 +108,56 @@ static void write_qrcode(FILE *output, QRcode *qr) {
+        if (!output)
+                output = stdout;
+
+        int fd, move_down = 2;
```
2. This value ```move_down = 2```, I also got after trying different initial values: 1, 3, 4. It was 2 that eliminated the blank line I had between the top border and the qr code, I need to understand that fully, so it gets standardized.

bluca replied:
add a comment above to explain why it is set as such

1awesomeJ commented:
3. After the function exits, I see "segmentation fault". 

These are the three required cleanups I see from my testing. I'll work on them, and any other ones you point out, so we can have this tool all set this week.

bluca commented on man/rules/meson.build:
```diff
@@ -898,6 +898,7 @@ manpages = [
   ''],
  ['systemd-boot-random-seed.service', '8', [], 'ENABLE_BOOTLOADER'],
  ['systemd-boot', '7', ['sd-boot'], 'ENABLE_BOOTLOADER'],
+ ['systemd-bsod', '8', [], ''],
```
this needs to be optional on HAVE_QRENCODE too

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,249 @@
+#include "main-func.h"
+#include "qrcode-util.h"
+#include "sd-id128.h"
+#include "sd-journal.h"
```
move the two sd-* includes in a separate block above this one

1awesomeJ replied:
ohh.. I just saw this review and the one about the man page.
I'll add them to my next push sir.

1awesomeJ commented:
> > 3. After the function exits, I see "segmentation fault".
> > 
> > These are the three required cleanups I see from my testing. I'll work on them, and any other ones you point out, so we can have this tool all set this week.
> 
> run it in gdb and see what the error is

Awesome!

1awesomeJ commented on src/shared/qrcode-util.c:
```diff
@@ -116,4 +185,46 @@ int print_qrcode(FILE *out, const char *header, const char *string) {
+                fprintf(out, "\n%s:\n\n", header);
+        }
+
+        qr_code_width = qr_code_height = qr->width + 8;
+        if (column + qr_code_width > tty_width)
+                column = tty_width - qr_code_width;
+
+        if (row + qr_code_height > tty_height)
+                row = tty_height - qr_code_height;
+
+        write_positioned_qrcode(out, qr, row, column);
```
This is what I've used to fix the tty overflow issue.
I tried it with 90%/90%, the QR code survived.

Is it good to go sir.

1awesomeJ commented:
This is me trying to start printing at 90%/90%
The QR code doesn't break, the column se
![IMG_20230726_134246_567](https://github.com/systemd/systemd/assets/109356020/2fd0ed88-0386-46ed-8675-e54f1520d433)
ems to be getting perfectly adjusted,
but the row goes too far up.
Looks like it's still something about the QR code printing two rows at once that I still need to figure out.

1awesomeJ commented:
```
(gdb) backtrace
#0  0x00007ffff78be4e1 in free () from /lib64/libc.so.6
#1  0x000000000040251e in freep (p=0x7fffffffe320) at ../src/basic/alloc-util.h:107
#2  0x000000000040440c in run (argc=1, argv=0x7fffffffe498) at ../src/journal/bsod.c:224
#3  0x0000000000404498 in main (argc=1, argv=0x7fffffffe498) at ../src/journal/bsod.c:250
(gdb) 
```

bluca commented:
> So it looks like all that's left now is for me to take another look at the double row printing of the qr code and make any necessary adjustments.
> 
> This tool should be all set today except you my mentors spot other areas to optimize.

How is this going?

1awesomeJ commented:
> > So it looks like all that's left now is for me to take another look at the double row printing of the qr code and make any necessary adjustments.
> > This tool should be all set today except you my mentors spot other areas to optimize.
> 
> How is this going?

It will be ready in about 2 hours sir.

1awesomeJ commented:
Okay, 
So with our default cursor position of 60% tty_height and 75% tty_width, there's no issue with the testing on my machine.

Now for a case of a screen that is 100 characters wide and 60 characters high.
If someone sets the qr code start position at 90%/90% (90, 54) our qr code won't fit since it's about 29characters square.

So what we're deciding in such a case is that a check should be made to see if the specified  column (90) + our qr_code width( 29) exceeds the tty width (100) then we won't mind the 90 that is being set, instead we'll use the start column as tty_width(100) - qr_code width(29)
So the start column becomes 71, not the ambitious 90. This works fine.


In thesame vein, for the row if (54 + 29) exceeds 60, then we don't want to use 54, instead we'll use 60 -29, which is 31.
This 31 however seems to appear as starting at 11 or so (too far up)

I read it's something about aspect ratio.

So these are what I have:

1awesomeJ commented:
![IMG_20230726_134246_567](https://github.com/systemd/systemd/assets/109356020/a7f02207-8d76-421f-b665-9ca4c3ce2bc9)
Setting row to tty_height - qrcode_height

1awesomeJ commented:
> ![IMG_20230726_134246_567](https://user-images.githubusercontent.com/109356020/256636705-a7f02207-8d76-421f-b665-9ca4c3ce2bc9.jpg) Setting row to tty_height - qrcode_height

Too far up.

1awesomeJ commented:
Setting row to tty_height - qrcode_height / 2.
Almost at the very bottom but it overflows
![IMG_20230727_193036_095](https://github.com/systemd/systemd/assets/109356020/79ed29c1-79f0-4285-970e-d6b46884e070)

1awesomeJ commented:
Setting row to tty_height - qrcode_height * 3U / 4U
Looks cool.
![IMG_20230727_193217_944](https://github.com/systemd/systemd/assets/109356020/6f9c1c81-aed1-49f9-a1d2-c80a9ba3ac55)

1awesomeJ commented:
Another issue I had was how to place the header string directly above the qr code.

That has noe been done.

So in my opinion, I think we're set sir.
![IMG_20230727_201737_586](https://github.com/systemd/systemd/assets/109356020/e5e55ee2-31db-44f4-8a8d-239a88483f93)

1awesomeJ commented on man/rules/meson.build:
```diff
@@ -898,6 +898,7 @@ manpages = [
   ''],
  ['systemd-boot-random-seed.service', '8', [], 'ENABLE_BOOTLOADER'],
  ['systemd-boot', '7', ['sd-boot'], 'ENABLE_BOOTLOADER'],
+ ['systemd-bsod', '8', [], 'HAVE_QRENCODE'],
```
Is this okay for the man page sir.

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,250 @@
+#include "sd-id128.h"
+#include "sd-journal.h"
+
+#include "alloc-util.h"
```
Are these well set sir?

1awesomeJ commented:
I have now pushed the updates sir.

1awesomeJ commented:
@bluca @poettering 
Is this ready for the green?

1awesomeJ commented:
I'll now move to varlink. I've been stuck at the JSON library usage. There seems to be lots of functions and data types.

1awesomeJ commented on src/shared/qrcode-util.c:
```diff
@@ -116,4 +185,49 @@ int print_qrcode(FILE *out, const char *header, const char *string) {
+                fprintf(out, "%s:\n\n", header);
+        }
+
+        if (column + qr_code_width > tty_width)
+                column = tty_width - qr_code_width;
+
+        /*Terminal characters are twice as high as they are wide so it's qr_code_height / 2,
+         *  our QR code prints an extra new line, so we have -1 as well */
+        if (row + qr_code_height > tty_height)
+                row = tty_height - (qr_code_height / 2 ) - 1;
```
@bluca, I found the explanation for the column issue, and I've added it in the comment here.

1awesomeJ commented:
That was sort of the only issue I had left.
So I woke up this early (4am) thinking. I realized my screen could fit as many as 164 characters wide, but only about 35 rows, and I was like "There's no way my 14-inch screen is 4.7 times as wide as it's high, it's not even twice as wide!". Then I realized that terminal characters are rectangular not squares. I suspected they're about twice as high as they're wide, so my 164:35 was more like 164:70. 

Then I read that terminals use monospaced fonts which occupy grids that are often 2:1 in aspect ratio.

1awesomeJ commented:
With this fix, when someone sets the qr code to start at 90%/90%,
The right adjustments are made and they only come this close to the very edge:
![IMG_20230728_051848_323](https://github.com/systemd/systemd/assets/109356020/efac98f5-1809-457a-94dc-b59628d40f1f)

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,266 @@
+        if (r < 0)
+                log_warning_errno(r, "QR code could not be printed, ignoring: %m");
+
+        r = set_terminal_cursor_position(fd, w.ws_row - 1, w.ws_col * 2U / 5U);
+        if (r < 0)
+                log_warning_errno(r, "Failed to move terminal cursor position, ignoring: %m");
+
+        r = loop_write(fd, "Press any key to exit...", strlen("Press any key to exit..."), /* do_poll = */false);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to write to terminal: %m");
```
we should inform the users that they can exit on key press.

mrc0mmand commented on meson.build:
```diff
@@ -4004,6 +4004,20 @@ public_programs += executable(
+                dependencies : [userspace,
+                                versiondep,
+                                libqrencode],
+                install_rpath : rootpkglibdir,
+                install_dir : rootlibexecdir,
```
You'll need to rebase on top of the latest main and change this part to:

```suggestion
                install_rpath : pkglibdir,
                install_dir : libexecdir,
```

to make the CIs happy again.

1awesomeJ replied:
> You'll need to rebase on top of the latest main and change this part to:
> 
> to make the CIs happy again.

Awesome 🤣🤣. Thank you sir!

bluca commented on man/systemd-bsod.xml:
```diff
@@ -0,0 +1,62 @@
+  "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
+<!-- SPDX-License-Identifier: LGPL-2.1-or-later -->
+
+<refentry id="systemd-bsod" xmlns:xi="http://www.w3.org/2001/XInclude">
```
this needs also `conditional='HAVE_QRENCODE'`

1awesomeJ replied:
> this needs also `conditional='HAVE_QRENCODE'`

ohh....

1awesomeJ replied:
Done sir!

bluca commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,266 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to move terminal cursor position, ignoring: %m");
+
+        r = loop_write(fd, "Press any key to exit...", strlen("Press any key to exit..."), /* do_poll = */false);
```
STRLEN() instead of strlen()

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -116,4 +185,49 @@ int print_qrcode(FILE *out, const char *header, const char *string) {
+                column = tty_width - qr_code_width;
+
+        /*Terminal characters are twice as high as they are wide so it's qr_code_height / 2,
+         *  our QR code prints an extra new line, so we have -1 as well */
```
extra space after `*` and missing space above after `/*`

1awesomeJ commented:
Right away on them sir.

1awesomeJ commented:
All done sir.
Meson rebuilds the entire project when i switch branches. It's why I delayed.

bluca commented on src/shared/qrcode-util.c:
```diff
@@ -75,7 +94,7 @@ static void write_qrcode(FILE *output, QRcode *qr) {
                                 fputs(UNICODE_LOWER_HALF_BLOCK, output);
                         else if (b)
-                                fputs(UNICODE_UPPER_HALF_BLOCK, output);
+                                 fputs(UNICODE_UPPER_HALF_BLOCK, output);
```
spurious space added

1awesomeJ replied:
> spurious space added

whow..

1awesomeJ commented:
Now, we need to register this service as a varlink API client that subscribes to journald to get notified when messages with log_level "emerg" are processed, so it can fetch the message and display it.

I've been stuck on implementing this subscribe feature in journald. Please could I get some high level guidance - not so high level, or some resources or docs recommendations? Thank you sirs.

bluca commented:
there's a merge conflict, needs a rebase

1awesomeJ commented:
okay sir.
I'll be home in 30minutes

1awesomeJ commented on src/journal/meson.build:
```diff
@@ -80,6 +80,18 @@ executables += [
                         threads,
                 ],
         },
+        executable_template + {
+                'name' : 'systemd-bsod',
+                'conditions' : ['HAVE_QRENCODE'],
+                'public' : true,
+                'sources' : files('bsod.c'),
+                'link_with' : libshared,
+                'dependencies' : [
+                        userspace,
+                        versiondep,
+                        libqrencode
+                ],
+        },
 ]
```
@bluca 
Did I leave anything out?

Sorry it's late sir.

AdrianVovk commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,266 @@
+
+        xsprintf(tty, "/dev/tty%d", free_vt + 1);
+
+        fd = open_terminal(tty, O_RDWR|O_NOCTTY|O_CLOEXEC);
```
You leak tty1's `fd` here

1awesomeJ replied:
> You leak tty1's `fd` here

😂😂

You remember me from that initrd guide you gave in May?
Thanks again for that.
Thanks for showing up here too.

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,266 @@
+        int r, free_vt = 0, original_vt = 0;
+        unsigned qr_code_start_row = 1, qr_code_start_column = 1;
+        char tty[STRLEN("/dev/tty") + DECIMAL_STR_MAX(int) + 1];
+        _cleanup_close_ int fd = -EBADF;
```
@AdrianVovk 
doesn't this ```_cleanup_close_ ``` take care of such leakages?

bluca replied:
it does not, use close_and_replace()

1awesomeJ replied:
> it does not, use close_and_replace()

ohh...
I already pushed a patch where I separated the fds.

I use rather use close_and_replace()?

1awesomeJ replied:
> > it does not, use close_and_replace()
> 
> ohh... I already pushed a patch where I separated the fds.
> 
> I should rather use close_and_replace()?

bluca replied:
yes

1awesomeJ replied:
> yes

On it sir.

1awesomeJ commented:
Looks like I'm hearing you say "tty1's fd should be closed before we assign fd to hold the descriptor of another tty" doesn't ```_cleanup_close_``` handle such scenarios?

1awesomeJ commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,267 @@
+
+        assert(message);
+
+        fd_tty1 = open_terminal("/dev/tty1", O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd_tty1 < 0)
+                return log_error_errno(fd, "Failed to open tty1: %m");
+
+        r = find_next_free_vt(fd_tty1, &free_vt, &original_vt);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find a free VT: %m");
+
+        xsprintf(tty, "/dev/tty%d", free_vt + 1);
+
+        fd = open_terminal(tty, O_RDWR|O_NOCTTY|O_CLOEXEC);
```
Like so?

AdrianVovk commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (tty1_fd < 0)
+                return log_error_errno(fd, "Failed to open tty1: %m");
+
+        close_and_replace(fd, tty1_fd);
```
Wrong place!

Basically where you put it now, you close fd (which is already -EBADF, so closing it does nothing). Then you replace fd with tty1_fd (and at the same time you make `tty1_fd = -EBADF`). Then later you overwrite `fd` with the real value you want, and so you're still leaking.

What you really want to do is something like this:

```c
_cleanup_close int fd = -EBADF;
int r;

fd = open_terminal("/dev/tty1", ...);
if (fd < 0) ...

...

r = open_terminal(tty, ...)
if (r < 0) ...
close_and_replace(fd, r) /* this closes fd (which is /dev/tty1) and replaces it with r (which is the tty you want) */
```

PS: also, both `fd` and `tty1_fd` have `_cleanup_close_` attached to them. Not sure if that's intended or not.

1awesomeJ commented:
Wow... awesome.
Thanks @AdrianVovk,
You explain like you're a college professor.
Are you?

Thanks again.

Let me fix.

1awesomeJ commented:
hey @AdrianVovk, I found you do actually have some business in college, although not actually as a professor.
I'd like to stay in touch if your schedule permits. I found a twitter profile which you don't seem to have used in a while. I followed you anyways.

Please let me know if your schedule has room for some sort of extra friend.
Thank you.

yuwata commented on src/basic/terminal-util.h:
```diff
@@ -68,6 +68,9 @@
 #define ANSI_HIGHLIGHT_YELLOW_FALLBACK           "\x1B[0;1;33m"
 #define ANSI_HIGHLIGHT_YELLOW_FALLBACK_UNDERLINE "\x1B[0;1;4;33m"
 
+/*Background colors */
```
Missing space after `/*`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+#include <errno.h>
+#include <getopt.h>
+#include <linux/vt.h>
+#include <sys/ioctl.h>
+#include <stdio.h>
```
Please sort them.

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+
+static void help(void) {
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the\n"
```
`filters` -> `Filter` (capitalize and drop 's')

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        printf("%s\n\n"
+               "filters the journal to fetch the first message from the\n"
+               "current boot with an emergency log level and displays it\n"
+               "as a string and as a QR code.\n\n"
```
`and a QR code`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get boot ID: %m");
+
+        xsprintf(boot_id_filter, "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));
+
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to add boot ID filter: %m");
+
```
Please use `add_match_this_boot()`.

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (ioctl(fd, VT_GETSTATE, &terminal_status) < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++)
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+
+        if (i >= sizeof(terminal_status.v_state) * 8)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTTY), "No free VT found: %m");
+
+        *ret_free_vt = i;
+        *ret_original_vt = terminal_status.v_active;
```
Inline the iterator and return earlier when found:
```c
        for (size_t i = 0; i < sizeof(terminal_status.v_state) * 8; i++)
                if ((terminal_status.v_state & (1 << i)) == 0) {
                        *ret_free_vt = i;
                        *ret_original_vt = terminal_status.v_active;
                        return 0;
                }

        return log_error_errno(SYNTHETIC_ERRNO(ENOTTY), "No free VT found: %m");


```

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (ioctl(fd, VT_ACTIVATE, free_vt + 1) < 0)
+                return log_error_errno(errno, "Failed to activate tty: %m");
+
+        r = loop_write(fd, ANSI_BACKGROUND_BLUE ANSI_HOME_CLEAR, STRLEN(ANSI_BACKGROUND_BLUE ANSI_HOME_CLEAR), /* do_poll = */ false);
```
`STRLEN()` -> `SIZE_MAX`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to move terminal cursor position, ignoring: %m");
+
+        r = loop_write(fd, "The current boot has failed!", STRLEN("The current boot has failed!"), /* do_poll = */false);
```
`STRLEN(...)` -> `SIZE_MAX`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to move terminal cursor position, ignoring: %m");
+
+        r = loop_write(fd, message, strlen(message), /* do_poll = */false);
```
Again, `strlen()` -> `SIZE_MAX`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+                switch (c) {
+
+                case 'h':
+                        help();
+                        return 0;
```
Please make `help()` return 0, then `return help();`

yuwata commented on src/journal/bsod.c:
```diff
@@ -0,0 +1,268 @@
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
```
Please add a link to the man page. See e.g. `help()` in battery-check.c.

yuwata commented on src/shared/qrcode-util.c:
```diff
@@ -48,6 +48,25 @@ static void print_border(FILE *output, unsigned width) {
         }
 }
 
+static void print_positioned_border(FILE *output, unsigned width, unsigned row, unsigned column) {
```
Hmm. `_positined` version of the functions should be merged with the original ones.
Instead of introducing new functions, 
1. please make `print_border()` and `print_qrcode()` take `row` and `column` (if they are `UINT_MAX` ignore the values),
2. rename `print_qrcode()` -> `print_qrcode_full()`,
3. then introduce an inline function in the header that takes `UINT_MAX` for `row` and `column`.

1awesomeJ commented:
Thank you so much @yuwata for the very detailed review.
Let me fix.

1awesomeJ commented:
@yuwata, I made all the requested changes.
Sorry I took long.

bluca commented:
needs a rebase on main due to conflicts

1awesomeJ commented:
> needs a rebase on main due to conflicts

Okay sir.
I'll fix in about an hour.

yuwata reviewed (approved):
LGTM. Let's go in this form at least now.
Future tasks are
- merge two qrcode logic more,
- add tests for bsod.service.

bluca commented:
When you rebase and push, please also amend the commit message, as this is not about PID1, it's a different tool

1awesomeJ commented:
> When you rebase and push, please also amend the commit message, as this is not about PID1, it's a different tool

Certainly sir.

1awesomeJ commented:
> LGTM. Let's go in this form at least now. Future tasks are
> 
> * merge two qrcode logic more,
> * add tests for bsod.service.

Alright sir. 
I hope "future" turns out to be next week or next month.

grepwood commented:
If you want BSOD just install Windows lol
